In [1]:
import numpy as np
import pandas as pd
from scipy import stats

# K32 - Phân tích dữ liệu thông minh

## Bài tập cá nhân 2

Họ tên học viên: Lê Nhựt Nam

MSHV: 22C11067

## Đề bài

Mô hình hồi quy giá nhà (PRICE tính bằng nghìn đô) phụ thuộc vào các biến SQFT (diện tích nhà) BEDRMS (số phòng ngủ) BATHS (số phòng tắm).

| $i$  |  PRICE | SQFT  | BEDRMS  | BATHS  |
|---|---|---|---|---|
| 1  | 199.9  | 1065  | 3  | 1.75  |
| 2  | 228  | 1254  | 3  | 2  |
| 3  | 235  | 1300  | 3  | 2  |
| 4  | 285  | 1577  | 4  | 2.5  |
| 5  | 239  | 1600  | 3  | 2  |
| 6  | 293  | 1750  | 4  | 2  |
| 7  | 285  | 1800  | 4  | 2.75  |
| 8  | 365  | 1870  | 4  | 2  |
| 9  | 295  | 1935  | 4  | 2.5  |
| 10  | 290  | 1948  | 4  | 2  |
| 11  | 385  | 2254  | 4  | 3  |
| 12  | 505  | 2600  | 3  | 2.5  |
| 13  | 425  | 2800  | 4  | 3  |
| 14  | 415  | 3000  | 4  | 3  |


a) Hồi quy mẫu

$$
\text{PRICE}_i = \beta_0 + \beta_1\text{SQFT}_i + \beta_2\text{BEDRMS}_i + \beta_3\text{BATHS}_i + \epsilon_i
$$

và giải thích ý nghĩa của các hệ số hồi quy?

b) Tính ước lượng điểm giá PRICE trung bình khi SQFT tăng thêm 300 và BEDRMS tăng thêm 1

c) Tính hệ số xác định và hệ số xác định điều chỉnh, giải thích ý nghĩa?

d) Tìm khoảng tin cậy đối với $\beta_1$, $\beta_2$, và $\beta_3$ với mức ý nghĩa 5%

e) Phải chăng cả ba biến đều không ảnh hưởng đến giá PRICE?

f) Dự báo giá trị trung bình và giá trị cá biệt về giá nhà trung bình khi SQRT = 2500, BDRMS = 3 và BATHS = 2?

Đọc dữ liệu

In [2]:
data = pd.read_csv('data.csv', ).reset_index().drop(['index', 'i'], axis=1)
data

PRICE  SQFT  BEDRMS  BATHS
0   199.9  1065       3   1.75
1   228.0  1254       3   2.00
2   235.0  1300       3   2.00
3   285.0  1577       4   2.50
4   239.0  1600       3   2.00
5   293.0  1750       4   2.00
6   285.0  1800       4   2.75
7   365.0  1870       4   2.00
8   295.0  1935       4   2.50
9   290.0  1948       4   2.00
10  385.0  2254       4   3.00
11  505.0  2600       3   2.50
12  425.0  2800       4   3.00
13  415.0  3000       4   3.00

Chuẩn bị dữ liệu

In [3]:
X = data[['SQFT', 'BEDRMS', 'BATHS']].values
X = np.hstack([np.ones((len(X),1)), X])
X

array([[1.000e+00, 1.065e+03, 3.000e+00, 1.750e+00],
       [1.000e+00, 1.254e+03, 3.000e+00, 2.000e+00],
       [1.000e+00, 1.300e+03, 3.000e+00, 2.000e+00],
       [1.000e+00, 1.577e+03, 4.000e+00, 2.500e+00],
       [1.000e+00, 1.600e+03, 3.000e+00, 2.000e+00],
       [1.000e+00, 1.750e+03, 4.000e+00, 2.000e+00],
       [1.000e+00, 1.800e+03, 4.000e+00, 2.750e+00],
       [1.000e+00, 1.870e+03, 4.000e+00, 2.000e+00],
       [1.000e+00, 1.935e+03, 4.000e+00, 2.500e+00],
       [1.000e+00, 1.948e+03, 4.000e+00, 2.000e+00],
       [1.000e+00, 2.254e+03, 4.000e+00, 3.000e+00],
       [1.000e+00, 2.600e+03, 3.000e+00, 2.500e+00],
       [1.000e+00, 2.800e+03, 4.000e+00, 3.000e+00],
       [1.000e+00, 3.000e+03, 4.000e+00, 3.000e+00]])

In [4]:
y = data['PRICE'].values
y

array([199.9, 228. , 235. , 285. , 239. , 293. , 285. , 365. , 295. ,
       290. , 385. , 505. , 425. , 415. ])

In [5]:
n = X.shape[0]
bar_ysquare = np.mean(y) ** 2

sum_y = np.sum(y)
sum_ysquare = np.sum(y) ** 2

sum_x1 = np.sum(data['SQFT'])
sum_x1square = np.sum(data['SQFT']) ** 2

sum_x2 = np.sum(data['BEDRMS'])
sum_x2square = np.sum(data['BEDRMS']) ** 2

sum_x2 = np.sum(data['BATHS'])
sum_x2square = np.sum(data['BATHS']) ** 2

sum_x1x2 = np.sum(data['SQFT'] * data['BEDRMS'])
sum_x1x3 = np.sum(data['SQFT'] * data['BATHS'])
sum_x2x3 = np.sum(data['BEDRMS'] * data['BATHS'])

sum_yx1 = np.sum(data['PRICE'] * data['SQFT'])
sum_yx2 = np.sum(data['PRICE'] * data['BEDRMS'])
sum_yx3 = np.sum(data['PRICE'] * data['BATHS'])

XTX = np.dot(X.T, X)
XTY = np.dot(X.T, y)

a) Hồi quy mẫu

$$
\text{PRICE}_i = \beta_0 + \beta_1\text{SQFT}_i + \beta_2\text{BEDRMS}_i + \beta_3\text{BATHS}_i + \epsilon_i
$$

và giải thích ý nghĩa của các hệ số hồi quy?

Trong mô hình hồi quy trên, PRICE là giá căn hộ tính theo nghìn dollars; SQFT thể hiện diện tích nhà; BEDRMS thể hiện số lượng phòng ngủ, và BATHS thể hiện số lượng nhà tắm.

Vì những đặc trưng này đều là những đặc tính tốt về một căn nhà, **ta kỳ vọng rằng các hệ số $\beta_1$, $\beta_2$, và $\beta_3$ có giá trị dương**.

Sử dụng công thức nghiệm hồi quy tuyến tính bằng phương pháp triển vọng cực đại, ta thu được giá trị của các hệ số trên công thức.

In [6]:
beta_ml = np.linalg.inv(X.T @ X) @ X.T @ y
beta_ml

# beta0: bias, khi không có bất kì yếu tố nào thì mức giá nhà tối thiểu là 129.06163452
# beta1: nếu diện tích nhà tăng 1 mét vuông thì giá nhà tăng 0.15479989
# beta2: nếu số lượng phòng ngủ tăng 1 phòng thì giá nhà giảm 21.58751918
# beta3: nếu số lượng nhà tắm tăng lên 1 thì giá nhà giảm 12.19275743
# Nếu cả diện tích nhà, phòng ngủ và số nhà tăm tăng như nào, thì giá căn hộ giảm 
# (0.15479989 - 21.58751918 - 12.19275743)

array([129.06163452,   0.15479989, -21.58751918, -12.19275743])

Kết quả cho thấy dấu đi kèm với hệ số BEDRMS và BATHS là không giống với kỳ vọng, chúng có giá trị âm. 

Ta tính toán giá trị phần dư độ lỗi

In [7]:
epsilon = y - X @ beta_ml
epsilon

array([ -7.92363962,  -6.03263041,  -6.15342558,  28.65090141,
       -48.59339407,   3.77414086,  -2.82128581,  57.19815347,
       -16.76746099, -29.87623834,  29.94775123,  68.70308968,
       -14.57299142, -55.53297041])

Một số kết luận:
- Thông thường, tăng thêm số lượng phòng ngủ hoặc nhà tắm thì giá trị căn nhà phải tăng lên. Nhưng ở đây, không thể hiện như vậy, điều đó là giá trị căn nhà giảm xuống. Liệu đây có phải một điều bất hợp lý hay không?

Xem xét giá trị riêng phần của từng biến lên giá của căn nhà. Bằng cách giữ nguyên diện tích sử dụng và số lượng phòng tắm. Kết quả trên cho ta thấy một điều rằng nếu tăng thêm một phòng ngủ, thì trung bình giá căn nhà giảm 21.58751918. 

Có thể nếu tăng thêm phòng ngủ, thì diện tích từng phòng sẽ nhỏ đi, điều này làm cho khách hàng không vừa lòng.

In [8]:
# Tính toán các thứ
TSS = np.sum(data['PRICE']**2) - X.shape[0] * np.mean(data['PRICE'])**2
ESS = beta_ml.T @ (X.T @ y) - X.shape[0] * np.mean(y) ** 2
RSS = TSS - ESS

In [9]:
# Phương sai
sigma_square = RSS / (n - 1)

# Hiệp phương sai
cov_beta = (sigma_square * np.linalg.inv(X.T @ X)).diagonal()

se_beta1 = np.sqrt(cov_beta[1])
se_beta2 = np.sqrt(cov_beta[2])
se_beta3 = np.sqrt(cov_beta[3])

varY0 = sigma_square * X[0].T @ np.linalg.inv(X.T @ X) @ X[0]
seY0 = np.sqrt(varY0)

varY0_hatY0 = varY0 + sigma_square
seY0_hatY0 = np.sqrt(varY0_hatY0)

b) Tính ước lượng điểm giá PRICE trung bình khi SQFT tăng thêm 300 và BEDRMS tăng thêm 1

Ta tính sự thay đổi của giá nhà nếu khi SQFT tăng thêm 300 và BEDRMS tăng thêm 1

In [10]:
delta_price = 0.15479989 * 300 - 21.58751918 * 1
delta_price

24.852447819999995

Về trung bình, giá căn hộ sẽ tăng 24.852447819999995 dollars.

Xử lý dữ liệu

In [11]:
Adj_X = np.array([data['SQFT'] + 300, data['BEDRMS'] + 1, data['BATHS']]).T
Adj_X = np.hstack([np.ones((len(Adj_X ),1)), Adj_X])
Adj_X

array([[1.000e+00, 1.365e+03, 4.000e+00, 1.750e+00],
       [1.000e+00, 1.554e+03, 4.000e+00, 2.000e+00],
       [1.000e+00, 1.600e+03, 4.000e+00, 2.000e+00],
       [1.000e+00, 1.877e+03, 5.000e+00, 2.500e+00],
       [1.000e+00, 1.900e+03, 4.000e+00, 2.000e+00],
       [1.000e+00, 2.050e+03, 5.000e+00, 2.000e+00],
       [1.000e+00, 2.100e+03, 5.000e+00, 2.750e+00],
       [1.000e+00, 2.170e+03, 5.000e+00, 2.000e+00],
       [1.000e+00, 2.235e+03, 5.000e+00, 2.500e+00],
       [1.000e+00, 2.248e+03, 5.000e+00, 2.000e+00],
       [1.000e+00, 2.554e+03, 5.000e+00, 3.000e+00],
       [1.000e+00, 2.900e+03, 4.000e+00, 2.500e+00],
       [1.000e+00, 3.100e+03, 5.000e+00, 3.000e+00],
       [1.000e+00, 3.300e+03, 5.000e+00, 3.000e+00]])

Tính toán lại mô hình hồi quy

In [12]:
adj_beta_ml = np.linalg.inv(X.T @ X) @ X.T @ y
adj_beta_ml

array([129.06163452,   0.15479989, -21.58751918, -12.19275743])

Đưa ra kết quả dự đoán

In [13]:
pred_price = Adj_X @ adj_beta_ml
df = pd.DataFrame({'Pred_PRICE': pred_price, 'Actual': y})
df

Pred_PRICE  Actual
0   232.676089   199.9
1   258.885080   228.0
2   266.005875   235.0
3   281.201548   285.0
4   312.445843   239.0
5   314.078308   293.0
6   312.673735   285.0
7   332.654296   365.0
8   336.619910   295.0
9   344.728688   290.0
10  379.904698   385.0
11  461.149360   505.0
12  464.425441   425.0
13  495.385420   415.0

c) Tính hệ số xác định và hệ số xác định điều chỉnh, giải thích ý nghĩa?

In [14]:
# Hệ số xác định R^2
# sự biến thiên của giá nhà từ SQFT, BEDRMS, BATHS, giải thích được 83%
R2 = ESS / TSS
R2

0.8359763585708145

In [15]:
# Hệ số xác định điều chỉnh adjusted-R^2
adjustedR2 = 1 - (1 - R2) * (len(y) - 1) / (len(y) - X.shape[1] - 1)
adjustedR2

0.7630769623800654

d) Tìm khoảng tin cậy đối với $\beta_1$, $\beta_2$, và $\beta_3$ với mức ý nghĩa 5%

In [16]:
# Các hệ số của mô hình hồi quy tuyến tính, beta1, beta2, và beta3
beta1 = beta_ml[1]
beta2 = beta_ml[2]
beta3 = beta_ml[3]

print("Coefficients of the linear model: beta_1 = {}, beta_2 = {}, beta_3 = {}".format(beta1, beta2, beta3))

Coefficients of the linear model: beta_1 = 0.1547998949655238, beta_2 = -21.587519177015714, beta_3 = -12.192757432630032


In [17]:
# Tính toán các giá trị khớp của y
pred_y = X @ beta_ml
pred_y

array([207.82363962, 234.03263041, 241.15342558, 256.34909859,
       287.59339407, 289.22585914, 287.82128581, 307.80184653,
       311.76746099, 319.87623834, 355.05224877, 436.29691032,
       439.57299142, 470.53297041])

In [18]:
# Tính MSE
mse = np.mean((y-pred_y)**2)
mse

1192.862041082679

In [19]:
# Giả định các hệ số của mô hình hồi quy thuộc phân phối t
# Tính t-statistics với độ tin cậy 95%, tức sử dụng mức ý nghĩa 5%
t_val = np.array([stats.t.ppf(0.95, X.shape[0] - 2)])
t_val

array([1.78228756])

In [20]:
# Giới hạn chặn trên và chặn dưới cho hệ số hồi quy của X ở độ tin cậy 95%
beta1_upper = beta1 + t_val * np.sqrt(mse) / np.sqrt(np.sum((X - np.mean(X))**2))
beta1_lower = beta1 - t_val * np.sqrt(mse) / np.sqrt(np.sum((X - np.mean(X))**2))
print("Upper bound beta_1 at 95% confidence: ", beta1_upper)
print("Estimation for beta_1: ", beta1)
print("Lower bound beta_1 at 95% confidence: ", beta1_lower)

Upper bound beta_1 at 95% confidence:  [0.16423242]
Estimation for beta_1:  0.1547998949655238
Lower bound beta_1 at 95% confidence:  [0.14536737]


In [21]:
# Giới hạn chặn trên và chặn dưới cho hệ số hồi quy của X ở độ tin cậy 95%
beta2_upper = beta2 + t_val * np.sqrt(mse) / np.sqrt(np.sum((X - np.mean(X))**2))
beta2_lower = beta2 - t_val * np.sqrt(mse) / np.sqrt(np.sum((X - np.mean(X))**2))
print("Upper bound beta_1 at 95% confidence: ", beta2_upper)
print("Estimation for beta_1: ", beta2)
print("Lower bound beta_1 at 95% confidence: ", beta2_lower)

Upper bound beta_1 at 95% confidence:  [-21.57808665]
Estimation for beta_1:  -21.587519177015714
Lower bound beta_1 at 95% confidence:  [-21.59695171]


In [22]:
# Giới hạn chặn trên và chặn dưới cho hệ số hồi quy của X ở độ tin cậy 95%
beta3_upper = beta3 + t_val * np.sqrt(mse) / np.sqrt(np.sum((X - np.mean(X))**2))
beta3_lower = beta3 - t_val * np.sqrt(mse) / np.sqrt(np.sum((X - np.mean(X))**2))
print("Upper bound beta_1 at 95% confidence: ", beta3_upper)
print("Estimation for beta_1: ", beta3)
print("Lower bound beta_1 at 95% confidence: ", beta3_lower)

Upper bound beta_1 at 95% confidence:  [-12.1833249]
Estimation for beta_1:  -12.192757432630032
Lower bound beta_1 at 95% confidence:  [-12.20218996]


e) Phải chăng cả ba biến đều không ảnh hưởng đến giá PRICE?

Tiến hành kiểm định giả thiết thống kê.

Giả định các hệ số hồi quy thuộc phân phối t. Ta sử dụng t-test

- Không có quan hệ: Không tồn tại quan hệ giữa biến phụ thuộc $y$ và biến độc lập $x_i$. Trong tình huống này, hệ số hồi quy $\beta_i$ bằng không. Điều này cho ta giả thiết rỗng/ không (null hypothesis) trong kiểm định hệ số hồi quy:

$$
H_0: \beta_i = 0
$$

- Có tồn tại quan hệ: Có tồn tại quan hệ giữa biến phụ thuộc $y$ và biến độc lập $x_i$. Trong tình huống này, hệ số hồi quy $\beta_i$ không bằng không. Điều này cho ta giả thiết thay thế (alternative hypothesis) trong kiểm định hệ số hồi quy:

$$
H_1: \beta_i \ne 0
$$

In [23]:
# Tính MSE
mse = np.mean((y-pred_y)**2)
mse

1192.862041082679

In [24]:
# Tính phương sai độ lỗi
var_err = mse*(np.linalg.inv(np.dot(X.T, X)).diagonal())
var_err

array([5.56961877e+03, 7.28708672e-04, 5.21846127e+02, 1.33611611e+03])

In [25]:
# Tính độ lệch chuẩn độ lỗi
std_err = np.sqrt(var_err)
std_err

array([7.46298786e+01, 2.69946045e-02, 2.28439516e+01, 3.65529221e+01])

In [26]:
# Tính t-values cho các hệ số
t_vals = beta_ml / std_err
t_vals

array([ 1.72935608,  5.73447538, -0.94499934, -0.33356451])

In [27]:
# Tính p-valus
p_values =[2*(1-stats.t.cdf(np.abs(i),(len(X)-len(X[0])))) for i in t_vals]
p_values

[0.11443375755441898,
 0.00018914452638507662,
 0.3669320942930625,
 0.7455970504342839]

In [28]:
# So sánh p-value với mức ý nghĩa
for idx, p_value in enumerate(p_values):
    if p_value > 0.02:
        print('The regression coefficient beta_{} of X is not statistically significant.'.format(idx))
    else:
        print('The regression coefficient beta_{} of X is statistically significant.'.format(idx))

The regression coefficient beta_0 of X is not statistically significant.
The regression coefficient beta_1 of X is statistically significant.
The regression coefficient beta_2 of X is not statistically significant.
The regression coefficient beta_3 of X is not statistically significant.


f) Dự báo giá trị trung bình và giá trị cá biệt về giá nhà trung bình khi SQRT = 2500, BDRMS = 3 và BATHS = 2?

In [29]:
SQRT = 2500
BDRMS = 3
BATHS = 2

# Tính giá trị dự đoán
y0_hat = 129.06163452 + beta1 * SQRT + beta2 * BDRMS + beta3 * BATHS

# t-value với mức ý nghĩa 5%
t_val = stats.t.ppf(0.95, X.shape[0] - 2)


# Khoảng dự báo cho giá trị trung bình E(Y/ SQRT = 2500, BDRMS = 3, BATHS = 2)
# y0_hat - t_val * SE(\hat{y_0}) <= E(Y/ SQRT = 2500, BDRMS = 3, BATHS = 2) <= y0_hat + t_val * SE(\hat{y_0})
mean_y0_hat_upper = y0_hat + t_val * seY0
mean_y0_hat_lower = y0_hat - t_val * seY0

print("Lower bound mean y0_hat at 95% confidence: ", mean_y0_hat_lower)
print("Upper bound mean y0_hat at 95% confidence: ", mean_y0_hat_upper)
print("------------------")

# Khoảng dự báo cho giá trị cá biệt Y0
individual_y0hat_upper = y0_hat + t_val * seY0_hatY0
individual_y0hat_lower = y0_hat - t_val * seY0_hatY0

print("Upper bound individual y0_hat at 95% confidence: ", individual_y0hat_upper)
print("Lower bound individual y0_hat at 95% confidence: ", individual_y0hat_lower)
print("------------------")

Lower bound mean y0_hat at 95% confidence:  393.4463570775016
Upper bound mean y0_hat at 95% confidence:  460.38024199750294
------------------
Upper bound individual y0_hat at 95% confidence:  499.0291550807673
Lower bound individual y0_hat at 95% confidence:  354.79744399423726
------------------
